### PR TITLE
Add Differential Diffusion to HunyuanDiT.

### PR DESCRIPTION
**Pull request by MnCSSJ4x**
# What does this PR do?
Adds [Differential Diffusion](https://arxiv.org/abs/2306.00950) to [HunyuanDIT](https://github.com/huggingface/diffusers/pull/8240). 

Fixes Partially #8924(HunyuanDiT Only) 

## Before submitting
- [x] Did you read the [contributor guideline](https://github.com/huggingface/diffusers/blob/main/CONTRIBUTING.md)?
- [x] Did you read our [philosophy doc](https://github.com/huggingface/diffusers/blob/main/PHILOSOPHY.md) (important for complex PRs)?

How to test:

## Gradient 

```python
import torch
from diffusers import FlowMatchEulerDiscreteScheduler
from diffusers.utils import load_image
from PIL import Image
from torchvision import transforms

from pipeline_hunyuandit_differential_img2img import (
    HunyuanDiTDifferentialImg2ImgPipeline,
)


pipe = HunyuanDiTDifferentialImg2ImgPipeline.from_pretrained(
    "Tencent-Hunyuan/HunyuanDiT-Diffusers", torch_dtype=torch.float16
).to("cuda")


source_image = load_image(
    "https://huggingface.co/datasets/OzzyGT/testing-resources/resolve/main/differential/20240329211129_4024911930.png"
)
map = load_image(
    "https://huggingface.co/datasets/OzzyGT/testing-resources/resolve/main/differential/gradient_mask_2.png"
)
prompt = "a green pear"
negative_prompt = "blurry"

image = pipe(
    prompt=prompt,
    negative_prompt=negative_prompt,
    image=source_image,
    num_inference_steps=28,
    guidance_scale=4.5,
    strength=1.0,
    map=map,
).images[0]
```

|  Gradient | Input | Output|
|--------------------------------------------|--------------------------------------------|--------------------------------------------|
| Gradient | Input | Output |

A colab notebook demonstrating all results can be found [here](https://colab.research.google.com/drive/1v44a5fpzyr4Ffr4v2XBQ7BajzG874N4P?usp=sharing). Depth Maps have also been added in the same colab. 

## Who can review?
@a-r-r-o-w @DN6

### CODE REVIEW

**Review by a-r-r-o-w — APPROVED** [review]
Thank you for working on this, looks good to me! I think we can merge this once you add your name and contrib to the community README file. Also, seems like at place of the code, the style guide is not followed. Normally, these could be fixed with `make style` if it was a pipeline in core diffusers. However, since this is a community pipeline, you can run styling with:
`ruff check examples/community/pipeline_hunyuandit_differential_img2img.py --fix`

**Review by a-r-r-o-w — APPROVED** [review]
Thank you for working on this, looks good to me! I think we can merge this once you add your name and contrib to the community README file. Also, seems like at place of the code, the style guide is not followed. Normally, these could be fixed with `make style` if it was a pipeline in core diffusers. However, since this is a community pipeline, you can run styling with:
`ruff check examples/community/pipeline_hunyuandit_differential_img2img.py --fix`

**Comment by MnCSSJ4x** [comment]
@a-r-r-o-w I have fixed the style issues and added the details to the markdown file. If all looks good, you can go ahead and merge this request.

**Comment by a-r-r-o-w** [comment]
@MnCSSJ4x Could you revert all the other changes apart from adding your name and contribution to community README? If you'd like to refactor, you can do it in a separate PR as it's out of scope for this one. Please keep the changes here limited

**Comment by MnCSSJ4x** [comment]
> @MnCSSJ4x Could you revert all the other changes apart from adding your name and contribution to community README? If you'd like to refactor, you can do it in a separate PR as it's out of scope for this one. Please keep the changes here limited

Sure. I'll try to revert in fix. I feel some tool might have auto refactored it.

**Comment by MnCSSJ4x** [comment]
@a-r-r-o-w Can you please check and let me know if its ok now ?

**Comment by a-r-r-o-w** [comment]
No problem, but there are still a large number of differences. Please run `git restore -s main examples/community/README.md` first, and then add your name to the table, and an example of how to run Hunyuan DiffDiff

**Comment by MnCSSJ4x** [comment]
@a-r-r-o-w Thanks for the command. It should be resolved now.

File: examples/community/README.md [hunk 1]
```diff
@@ -1641,18 +1642,18 @@ from io import BytesIO
...
 scheduler = DDIMScheduler.from_pretrained("stabilityai/stable-diffusion-2-1",
                                             subfolder="scheduler")
 
 
-pipe = DiffusionPipeline.from_pretrained("stabilityai/stable-diffusion-2-1",
-                                            custom_pipeline="stable_diffusion_tensorrt_img2img",
-                                            variant='fp16',
-                                            torch_dtype=torch.float16,
-                                            scheduler=scheduler,)
+pipe = StableDiffusionImg2ImgPipeline.from_pretrained("stabilityai/stable-diffusion-2-1",
```
Comment (a-r-r-o-w) [inline examples/community/README.md:1652]: Please revert these changes.

File: examples/community/README.md [hunk 2]
```diff
@@ -1641,18 +1642,18 @@ from io import BytesIO
 from PIL import Image
 import torch
 from diffusers import DDIMScheduler
-from diffusers import DiffusionPipeline
+from diffusers.pipelines.stable_diffusion import StableDiffusionImg2ImgPipeline
```
Comment (a-r-r-o-w) [inline examples/community/README.md:1645]: Please revert this change.

File: examples/community/pipeline_hunyuandit_differential_img2img.py [hunk 3]
```diff
@@ -0,0 +1,1239 @@
...
+    import torch_xla.core.xla_model as xm
+
+    XLA_AVAILABLE = True
+else:
+    XLA_AVAILABLE = False
+
+
+logger = logging.get_logger(__name__)  # pylint: disable=invalid-name
+
+EXAMPLE_DOC_STRING = """
```
Comment (a-r-r-o-w) [inline examples/community/pipeline_hunyuandit_differential_img2img.py:57]: Could you update the example with Hunyuan DiffDiff example?

**Comment by a-r-r-o-w** [comment]
@MnCSSJ4x Looking good implementation-wise. The quality tests seems to be failing. Could you run `make style` and ensure that the tests pass locally too.

**Comment by MnCSSJ4x** [comment]
> @MnCSSJ4x Looking good implementation-wise. The quality tests seems to be failing. Could you run `make style` and ensure that the tests pass locally too.


Upon running ``make style`` I get the following and am unable to track back where exactly things are going wrong or if I introduced this issues. 
```python 
ruff check examples scripts src tests utils benchmarks setup.py --fix
src/diffusers/configuration_utils.py:679:16: E721 Use `is` and `is not` for type comparisons, or `isinstance()` for isinstance checks
    |
677 |             if field.name in self._flax_internal_args:
678 |                 continue
679 |             if type(field.default) == dataclasses._MISSING_TYPE:
    |                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ E721
680 |                 default_kwargs[field.name] = None
681 |             else:
    |

tests/models/test_modeling_common.py:338:20: E721 Use `is` and `is not` for type comparisons, or `isinstance()` for isinstance checks
    |
337 |         model.set_default_attn_processor()
338 |         assert all(type(proc) == AttnProcessorNPU for proc in model.attn_processors.values())
    |                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ E721
339 |         with torch.no_grad():
340 |             if self.forward_requires_fresh_args:
    |

tests/models/test_modeling_common.py:346:20: E721 Use `is` and `is not` for type comparisons, or `isinstance()` for isinstance checks
    |
345 |         model.enable_npu_flash_attention()
346 |         assert all(type(proc) == AttnProcessorNPU for proc in model.attn_processors.values())
    |                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ E721
347 |         with torch.no_grad():
348 |             if self.forward_requires_fresh_args:
    |

tests/models/test_modeling_common.py:354:20: E721 Use `is` and `is not` for type comparisons, or `isinstance()` for isinstance checks
    |
353 |         model.set_attn_processor(AttnProcessorNPU())
354 |         assert all(type(proc) == AttnProcessorNPU for proc in model.attn_processors.values())
    |                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ E721
355 |         with torch.no_grad():
356 |             if self.forward_requires_fresh_args:
    |

tests/models/test_modeling_common.py:389:20: E721 Use `is` and `is not` for type comparisons, or `isinstance()` for isinstance checks
    |
388 |         model.set_default_attn_processor()
389 |         assert all(type(proc) == AttnProcessor for proc in model.attn_processors.values())
    |                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^ E721
390 |         with torch.no_grad():
391 |             if self.forward_requires_fresh_args:
    |

tests/models/test_modeling_common.py:397:20: E721 Use `is` and `is not` for type comparisons, or `isinstance()` for isinstance checks
    |
396 |         model.enable_xformers_memory_efficient_attention()
397 |         assert all(type(proc) == XFormersAttnProcessor for proc in model.attn_processors.values())
    |                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ E721
398 |         with torch.no_grad():
399 |             if self.forward_requires_fresh_args:
    |

tests/models/test_modeling_common.py:405:20: E721 Use `is` and `is not` for type comparisons, or `isinstance()` for isinstance checks
    |
404 |         model.set_attn_processor(XFormersAttnProcessor())
405 |         assert all(type(proc) == XFormersAttnProcessor for proc in model.attn_processors.values())
    |                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ E721
406 |         with torch.no_grad():
407 |             if self.forward_requires_fresh_args:
    |

tests/models/test_modeling_common.py:433:20: E721 Use `is` and `is not` for type comparisons, or `isinstance()` for isinstance checks
    |
431 |             return
432 | 
433 |         assert all(type(proc) == AttnProcessor2_0 for proc in model.attn_processors.values())
    |                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ E721
434 |         with torch.no_grad():
435 |             if self.forward_requires_fresh_args:
    |

tests/models/test_modeling_common.py:441:20: E721 Use `is` and `is not` for type comparisons, or `isinstance()` for isinstance checks
    |
440 |         model.set_default_attn_processor()
441 |         assert all(type(proc) == AttnProcessor for proc in model.attn_processors.values())
    |                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^ E721
442 |         with torch.no_grad():
443 |             if self.forward_requires_fresh_args:
    |

tests/models/test_modeling_common.py:449:20: E721 Use `is` and `is not` for type comparisons, or `isinstance()` for isinstance checks
    |
448 |         model.set_attn_processor(AttnProcessor2_0())
449 |         assert all(type(proc) == AttnProcessor2_0 for proc in model.attn_processors.values())
    |                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ E721
450 |         with torch.no_grad():
451 |             if self.forward_requires_fresh_args:
    |

tests/models/test_modeling_common.py:457:20: E721 Use `is` and `is not` for type comparisons, or `isinstance()` for isinstance checks
    |
456 |         model.set_attn_processor(AttnProcessor())
457 |         assert all(type(proc) == AttnProcessor for proc in model.attn_processors.values())
    |                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^ E721
458 |         with torch.no_grad():
459 |             if self.forward_requires_fresh_args:
    |

tests/pipelines/controlnet/test_controlnet_sdxl.py:1022:16: E721 Use `is` and `is not` for type comparisons, or `isinstance()` for isinstance checks
     |
1021 |         controlnet = ControlNetModel.from_unet(unet, conditioning_channels=4)
1022 |         assert type(controlnet.mid_block) == UNetMidBlock2D
     |                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ E721
1023 |         assert controlnet.conditioning_channels == 4
     |

tests/pipelines/test_pipelines_common.py:777:21: E721 Use `is` and `is not` for type comparisons, or `isinstance()` for isinstance checks
    |
775 |             if hasattr(component, "attn_processors"):
776 |                 assert all(
777 |                     type(proc) == AttnProcessor for proc in component.attn_processors.values()
    |                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^ E721
778 |                 ), "`from_pipe` changed the attention processor in original pipeline."
    |

tests/schedulers/test_schedulers.py:827:16: E721 Use `is` and `is not` for type comparisons, or `isinstance()` for isinstance checks
    |
825 |         scheduler_loaded = DDIMScheduler.from_pretrained(f"{USER}/{self.repo_id}")
826 | 
827 |         assert type(scheduler) == type(scheduler_loaded)
    |                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ E721
828 | 
829 |         # Reset repo
    |

tests/schedulers/test_schedulers.py:838:16: E721 Use `is` and `is not` for type comparisons, or `isinstance()` for isinstance checks
    |
836 |         scheduler_loaded = DDIMScheduler.from_pretrained(f"{USER}/{self.repo_id}")
837 | 
838 |         assert type(scheduler) == type(scheduler_loaded)
    |                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ E721
839 | 
840 |         # Reset repo
    |

tests/schedulers/test_schedulers.py:854:16: E721 Use `is` and `is not` for type comparisons, or `isinstance()` for isinstance checks
    |
852 |         scheduler_loaded = DDIMScheduler.from_pretrained(self.org_repo_id)
853 | 
854 |         assert type(scheduler) == type(scheduler_loaded)
    |                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ E721
855 | 
856 |         # Reset repo
    |

tests/schedulers/test_schedulers.py:865:16: E721 Use `is` and `is not` for type comparisons, or `isinstance()` for isinstance checks
    |
863 |         scheduler_loaded = DDIMScheduler.from_pretrained(self.org_repo_id)
864 | 
865 |         assert type(scheduler) == type(scheduler_loaded)
    |                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ E721
866 | 
867 |         # Reset repo
    |

Found 17 errors.
make: *** [style] Error 1 
```

**Comment by a-r-r-o-w** [comment]
Is your ruff version same as the one in our setup.py? I remember seeing something in the past due to incompatible ruff versions

**Comment by MnCSSJ4x** [comment]
> Is your ruff version same as the one in our setup.py? I remember seeing something in the past due to incompatible ruff versions

Hi, yes the version was different. Fixed it and ran the command ``make style``, I think the formatting was fixed and I pushed the same. 

However, Got some error text 
```
ruff check examples scripts src tests utils benchmarks setup.py --fix
ruff format examples scripts src tests utils benchmarks setup.py
1 file reformatted, 1036 files left unchanged
doc-builder style src/diffusers docs/source --max_len 119
make: doc-builder: No such file or directory
make: *** [style] Error 1
```

**Review by a-r-r-o-w — APPROVED** [review]
Thank you for your contribution and bearing with our reviews! This is a very strong good-first-issue finish :tada:
